// @nciocpl/ncids-js/lib/cjs/components/usa-footer/utils/footer-back-to-top.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NCIBackToTop = void 0;
class NCIBackToTop {
    constructor(element) {
        this.scrollListener = () => this.handleScroll();
        this.element = element;
        this.initialize();
    }
    handleScroll() {
        const y = window.scrollY;
        if (y > 0) {
            this.element.classList.remove('hide');
            this.element.classList.add('show');
        }
        else {
            this.element.classList.remove('show');
            this.element.classList.add('hide');
        }
    }
    unregister() {
        this.element.classList.remove('show', 'hide', 'usa-footer__nci-return-to-top');
        this.element.classList.add('grid-container', 'usa-footer__return-to-top');
    }
    initialize() {
        this.element.classList.remove('grid-container', 'usa-footer__return-to-top');
        this.element.classList.add('usa-footer__nci-return-to-top', 'hide');
        window.addEventListener('scroll', this.scrollListener, false);
    }
}
exports.NCIBackToTop = NCIBackToTop;
