// @nciocpl/ncids-js/lib/esm/components/usa-site-alert/nci-site-alert.component.js
import { SiteAlertCloseButton } from './utils/site-alert-close-button.component';
import { SiteAlertCollapse } from './utils/site-alert-collapse.component';
export class NCISiteAlert {
    constructor(element, options) {
        this.element = element;
        this.options = Object.assign(Object.assign({}, NCISiteAlert.optionDefaults), options);
        const existingComponent = NCISiteAlert._components.get(this.element);
        if (existingComponent) {
            existingComponent.unregister();
        }
        NCISiteAlert._components.set(this.element, this);
        this.initialize();
    }
    static create(element, options) {
        if (!(element instanceof HTMLElement)) {
            throw 'Element is not an HTMLElement';
        }
        return this._components.get(element) || new this(element, options);
    }
    static autoInit() {
        document.addEventListener('DOMContentLoaded', () => {
            const siteAlerts = Array.from(document.getElementsByClassName('usa-site-alert'));
            siteAlerts.forEach((siteAlert) => {
                var _a;
                const alert = siteAlert;
                const closeable = ((_a = alert.dataset.siteAlertClosable) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === 'true';
                NCISiteAlert.create(alert, { closeable });
            });
        });
    }
    unregister() {
        if (this.collapse) {
            this.collapse.unregister();
        }
        if (this.closeButton) {
            this.closeButton.unregister();
        }
        NCISiteAlert._components.delete(this.element);
    }
    initialize() {
        this.createCloseButton();
        this.createCollapse();
    }
    createCloseButton() {
        if (this.options.closeable && this.element.id) {
            this.closeButton = new SiteAlertCloseButton(this.element, this.options);
        }
    }
    createCollapse() {
        if (this.element.classList.contains('usa-site-alert--nci-standard') &&
            this.element.id) {
            this.collapse = new SiteAlertCollapse(this.element, this.options);
        }
    }
}
NCISiteAlert.optionDefaults = {
    closeable: false,
    closeAriaLabel: 'Dismiss alert',
    closeButtonClass: 'usa-alert__nci-button--close',
    closeEventListenerLabel: 'usa-site-alert:close-button',
    closeCookiePath: '/',
    collapseAriaLabel: 'Toggle alert message',
    collapseButtonClass: 'usa-alert__nci-button--toggle',
    collapseCookiePath: '/',
    collapseEventListenerLabel: 'usa-site-alert:content',
};
NCISiteAlert._components = new Map();
